torchvision/prototype/datasets/_builtin/voc.py

```diff
@@ -12,48 +13,51 @@
     LineReader,
 )
 from torchvision.datasets import VOCDetection
-from torchvision.prototype.datasets.utils import (
-    Dataset,
-    DatasetConfig,
-    DatasetInfo,
-    HttpResource,
-    OnlineResource,
-)
+from torchvision.prototype.datasets.utils import DatasetInfo, OnlineResource, HttpResource, Dataset2
 from torchvision.prototype.datasets.utils._internal import (
     path_accessor,
     getitem,
     INFINITE_BUFFER_SIZE,
     path_comparator,
     hint_sharding,
     hint_shuffling,
+    BUILTIN_DIR,
 )
 from torchvision.prototype.features import BoundingBox, Label, EncodedImage
 
+from .._api import register_dataset, register_info
 
-class VOCDatasetInfo(DatasetInfo):
-    def __init__(self, *args: Any, **kwargs: Any):
-        super().__init__(*args, **kwargs)
-        self._configs = tuple(config for config in self._configs if config.split != "test" or config.year == "2007")
+NAME = "voc"
 
-    def make_config(self, **options: Any) -> DatasetConfig:
-        config = super().make_config(**options)
-        if config.split == "test" and config.year != "2007":
-            raise ValueError("`split='test'` is only available for `year='2007'`")
+CATEGORIES, *_ = zip(*DatasetInfo.read_categories_file(BUILTIN_DIR / f"{NAME}.categories"))
 
-        return config
 
+@register_info(NAME)
+def _info() -> Dict[str, Any]:
+    return dict(categories=CATEGORIES)
 
-class VOC(Dataset):
-    def _make_info(self) -> DatasetInfo:
-        return VOCDatasetInfo(
-            "voc",
-            homepage="http://host.robots.ox.ac.uk/pascal/VOC/",
-            valid_options=dict(
-                split=("train", "val", "trainval", "test"),
-                year=("2012", "2007", "2008", "2009", "2010", "2011"),
-                task=("detection", "segmentation"),
-            ),
-        )
+
+@register_dataset(NAME)
+class VOC(Dataset2):
+    def __init__(
+        self,
+        root: Union[str, pathlib.Path],
+        *,
+        split: str = "train",
+        year: str = "2012",
+        task: str = "detection",
+        **kwargs: Any,
+    ) -> None:
+        self._year = self._verify_str_arg(year, "year", ("2007", "2008", "2009", "2010", "2011", "2012"))
+        if split == "test" and year != "2007":
+            raise ValueError("`split='test'` is only available for `year='2007'`")
+        else:
+            self._split = self._verify_str_arg(split, "split", ("train", "val", "trainval", "test"))
+        self._task = self._verify_str_arg(task, "task", ("detection", "segmentation"))
+
+        self._categories: List[str] = _info()["categories"]
```

Comment:
Are we going to need to set types for these every single time?

Reply:
Nope. I've copy-pasted them from the imagenet implementation, but it seems we also don't need them there anymore. I'll remove.